sdk/eventhub/Azure.Messaging.EventHubs/src/EventHubConsumer.cs

```diff
@@ -237,5 +348,260 @@ protected EventHubConsumer()
+        ///
+        /// <returns>A task to be resolved on when the operation has completed.</returns>
+        ///
+        private Task StartBackgroundChannelPublishingAsync(CancellationToken cancellationToken) =>
+            Task.Run(async () =>
+            {
+                var failedAttemptCount = 0;
+                var receivedItems = default(IEnumerable<EventData>);
+                var retryDelay = default(TimeSpan?);
+                var activeException = default(Exception);
+
+                while (!cancellationToken.IsCancellationRequested)
+                {
+                    try
+                    {
+                        // Receive items in batches and then write them to the subscribed channels.  The channels will naturally
+                        // block if they reach their maximum queue size, so there is no need to throttle publishing.
+
+                        receivedItems = await InnerConsumer.ReceiveAsync(BackgroundPublishReceiveBatchSize, BackgroundPublishingWaitTime, cancellationToken).ConfigureAwait(false);
+
+                        await Task.WhenAll(
+                            receivedItems.SelectMany(item => ActiveChannels.Select(pair => pair.Value.Writer.WriteAsync(item, cancellationToken).AsTask()))
+                        ).ConfigureAwait(false);
+
+                        failedAttemptCount = 0;
+                    }
+                    catch (TaskCanceledException ex)
+                    {
+                        // In the case that a task was canceled, if cancellation was requested, then publishing should
+                        // is already terminating.  Otherwise, something unexpected canceled the operation and it should
+                        // be treated as an exception to ensure that the channels are marked final and consumers are made
+                        // aware.
+
+                        activeException = (cancellationToken.IsCancellationRequested) ? null : ex;
+                        break;
+                    }
+                    catch (ConsumerDisconnectedException ex)
+                    {
+                        // If the consumer was disconnected, it is known to be unrecoverable; do not offer the chance to retry.
+
+                        activeException = ex;
+                        break;
+                    }
+                    catch (Exception ex) when
+                        (ex is OutOfMemoryException
+                        || ex is StackOverflowException
+                        || ex is ThreadAbortException)
+                    {
+                        // These exceptions are known to be unrecoverable and there should be
+                        // no attempts at further processing.  The environment is in a bad state and
+                        // should fail.
+
+                        Environment.FailFast(Resources.UnrecoverableException, ex);
```

Comment:
Don't fail fast on your own. StackOverflowException ThreadAbortException only kill the thread.

Reply:
This was another carry over from track one; seemed odd to me but I didn't have a better idea.   Since there's no chance to recover and performing any of the clean up operations are risky, what would you propose doing?   Just throwing to bubble things up, or some other signal?

Reply:
Chatted a bit about this on the side and, for now, went with the "try to do minimal cleanup with as few allocations as possible and bubble" approach.   Most of those are unlikely to be restricted to the background thread and will just short-circuit on failure and bubble.

If you've got a better approach in mind, I'd love to hear it.

Reply:
My approach is to let things bubble up and only run normal finally statements (usings, locks, etc.)